fix cinder detach problems

zetaab opened the pull request:
**What this PR does / why we need it**: We have currently huge problems in cinder volume detach. This PR tries to fix these issues.

**Which issue(s) this PR fixes**:
Fixes #50004 
Fixes #57497

**Special notes for your reviewer**: 
**Release note**:

```release-note
openstack cinder detach problem is fixed if nova is shutdowned
```

dims commented:
/ok-to-test

dims commented:
/assign @FengyunPan

zetaab commented:
actually this attachdisk function is not called nowadays right after it needs the disk

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -317,8 +317,16 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
 			return volume.ID, nil
+		} else {
+			// otherwise the disk is in incorrect compute and it needs detach
+			glog.V(2).Infof("Disk %s is attached to a different compute: %s, attempting detach", volumeID, volume.AttachedServerId)
```
So this patch unequivocally performs detach from wherever the volume was previously detached in order for attach to succeed? I do not think this is correct approach and might cause race conditions. 

What I would suggest instead is to take approach we took for fixing EBS volumes and raise `volumeutil.NewDanglingError` if volume is attached to a compute node and it shouldn't be attached there - https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L2251

What that will do is, add the volume back into actual state of world of attach/detach controller and if there is no pod that needs the volume, then the volume will be correctly detached.

zetaab replied:
sounds like a plan

gnufied replied:
You are going to update this PR as aforementioned heuristic right?

zetaab replied:
Yep, i will update when i have time which is not now

dims replied:
cool. let me place a hold on this PR for now

/hold

zetaab committed:
use danglingerror

add getNodeNameByID and use volume.AttachedDevice as devicepath

use uppercase functionname

do not delete automatically nodes if node is shutdowned in openstack

do not delete node

fix gofmt

fix cinder detach if instance is not in active state

fix gofmt

zetaab commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -317,8 +317,20 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
+
+			//nodeName := mapServerToNodeName(attachedInstance)
+
+			// should we use os.GetDevicePathBySerialId(volumeID) or volume.AttachedDevice see issue #33128
```
I am now thinking could #50004 happen because of #33128? If devices are detached by devicepath, that might cause that problem that drives are not detached at all?

Correct way of doing danglingerror is maybe use volume.AttachDevice, but if that returns incorrect information we have have quite big problem. Instead if we insert "/dev/disk/by-id/", id to danglingerror, it might mean that it does not work at all?

Then second problem is that in openstack we do not have function like getinstancebyid. Where I can get nodename using instanceid without querying openstack api? That is of course the last option, but I would not like go there.

zetaab replied:
@kiall because you fixed that devicepath thing, could you check comment above ^.

gnufied replied:
>  Where I can get nodename using instanceid without querying openstack api? That is of course the last option, but I would not like go there.

Why would you not want to go there? It seems pretty trivial to get `nodeName` from instance-id using API. Something like:

```
server, err := servers.Get(i.compute, instanceID).Extract()
```

cc @NickrenREN

zetaab replied:
Well, if there is option to use existing data or load data from api always. My choice is use existing data. Like in this case nodename is quite static info - we could cache it.

But if there isnt option to use existing data then i must use api.

Still question is about that devicepath things. From my perspective it might be that these cinder detach issues might be related to incorrect paths. Cinder in openstack says like /dev/vdb but then when you go to machine it is located in /dev/vdc. I have seen problems like this in our openstack. However, i think we should here make kube trust cinder devicepath. Then openstack guys should fix their side. What you think?

gnufied replied:
I do not think detaching is related to device path problem because volume detach code doesn't appear to be using device path at all. 

If somehow detach is affected by incorrect device path, then it could be because umount is not working properly. I am yet to properly look into bugs linked in this PR to verify if that was the case.

gnufied commented:
@zetaab have you verified that this PR fixes problems you noticed in some sort of test cluster?

zetaab commented:
@gnufied I have tried to reproduce that issue with latest kube master, with and without this fix. Still the code is not going there. I have been wondering is the problem now that I am testing in very small cluster but no idea. Actually I have made ticket to fix this in openshift, and I think https://github.com/kubernetes/kubernetes/pull/52221 this will help to that problem (hopefully). According to logfiles when this cinder problem happens we see lots of that happening. So I am not sure is this PR needed?

xsgordon commented:
/sig openstack
/sig storage

FengyunPan commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -319,6 +319,22 @@ func mapNodeNameToServerName(nodeName types.NodeName) string {
+		return nodeName, err
+	}
+
+	server, serr := servers.Get(client, instanceID).Extract()
```
For go style, please use err.

FengyunPan commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -319,6 +319,22 @@ func mapNodeNameToServerName(nodeName types.NodeName) string {
+		return nodeName, err
+	}
+
+	server, serr := servers.Get(client, instanceID).Extract()
```
For go style, please use err.

gnufied replied:
I think it is fine. Go's variable shadowing rules can be awkward sometimes. @zetaab is being explicit about variable names - it doesn't seem like a big deal.

FengyunPan commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -317,8 +319,18 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
 			glog.V(4).Infof("Disk %s is already attached to instance %s", volumeID, instanceID)
 			return volume.ID, nil
+		} else {
+			nodeName, ierr := os.GetNodeNameByID(instanceID)
```
For go style, please use 'err'.

FengyunPan reviewed (commented):
@zetaab  Thank for your PR, it seems a bug, so have you tested this PR?

zetaab commented:
@FengyunPan could you check this PR. 

Before this PR if we shutdowned node in openstack and that node contained pod with volume. Then there was no way to get it working without releasing cinder from openstack gui or api. However, now after this PR if node is shutdowned, pod cannot still move to another node because of drive but after I start that node again the pod restarts itself and continues working. So we have even some progress here.

So what happens still if we have situation that I will just kill one node from openstack:

```
% kubectl get pods
NAME                               READY     STATUS              RESTARTS   AGE
nginx-deployment-5fc77998d-ffwpp   0/1       ContainerCreating   0          6m
nginx-deployment-5fc77998d-mtx6j   1/1       Unknown             0          8m
```

We have one pod "starting" and one pod which state is unknown and drive is attached in that shutdowned node. The problem here is that rows https://github.com/zetaab/kubernetes/blob/fixvolumeattached/pkg/cloudprovider/providers/openstack/openstack_volumes.go#L323-L332 is not called at all if there is situation like that. How kube should behave if we have node in notready situation? Start evacuation pods to another nodes? It looks like it starts that but it do not detach drives in that case.

zetaab commented:
I think now this PR is like what was mentioned by @gnufied. However, it does not detach volume like it should?

How I test this:
1. make pod with pvc to some node
2. shutdown that node from openstack api
3. after sometime pod is moved to another node

I can see in events

```
Events:
  Type     Reason                 Age              From                        Message
  ----     ------                 ----             ----                        -------
  Normal   Scheduled              3m               default-scheduler           Successfully assigned nginx-deployment-5fc77998d-gfws9 to kube-dev-node-1-0
  Normal   SuccessfulMountVolume  3m               kubelet, kube-dev-node-1-0  MountVolume.SetUp succeeded for volume "default-token-v2vhf"
  Warning  FailedAttachVolume     3m               attachdetach-controller     Multi-Attach error for volume "pvc-e46022b2-ea91-11e7-bfcf-fa163ee4a110" Volume is already exclusively attached to one node and can't be attached to another
  Warning  FailedMount            1m (x3 over 3m)  attachdetach-controller     AttachVolume.Attach failed for volume "pvc-e46022b2-ea91-11e7-bfcf-fa163ee4a110" : disk 0c12971f-da26-44a7-9333-1dba7e3f7be6 path /dev/vdb is attached to a different instance (12db4669-6145-449e-99b7-453031b71bb0)
  Warning  FailedMount            1m               kubelet, kube-dev-node-1-0  Unable to mount volumes for pod "nginx-deployment-5fc77998d-gfws9_default(25d323df-ea92-11e7-bfcf-fa163ee4a110)": timeout expired waiting for volumes to attach/mount for pod "default"/"nginx-deployment-5fc77998d-gfws9". list of unattached/unmounted volumes=[vol]
```

```
{"log":"I1226 23:12:02.694561       1 openstack_volumes.go:332] volume 0c12971f-da26-44a7-9333-1dba7e3f7be6 is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2017-12-26T23:12:02.695265124Z"}
{"log":"I1226 23:12:35.542536       1 openstack_volumes.go:332] volume 0c12971f-da26-44a7-9333-1dba7e3f7be6 is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2017-12-26T23:12:35.543368776Z"}
{"log":"I1226 23:13:40.519154       1 openstack_volumes.go:332] volume 0c12971f-da26-44a7-9333-1dba7e3f7be6 is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2017-12-26T23:13:40.520194944Z"}
{"log":"I1226 23:15:44.597998       1 openstack_volumes.go:332] volume 0c12971f-da26-44a7-9333-1dba7e3f7be6 is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2017-12-26T23:15:44.598172774Z"}
{"log":"I1226 23:17:48.082175       1 openstack_volumes.go:332] volume 0c12971f-da26-44a7-9333-1dba7e3f7be6 is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2017-12-26T23:17:48.082387008Z"}
```

So I am raising now NewDanglingError, but it looks like that its not detaching the volume?! The cinder volume is still in node-1-1 which is shutdowned and cannot be found from cluster.

```
{"log":"I1226 23:21:55.189745       1 openstack_volumes.go:332] volume 0c12971f-da26-44a7-9333-1dba7e3f7be6 is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2017-12-26T23:21:55.189918768Z"}
{"log":"I1226 23:21:55.189860       1 attacher.go:159] Attach volume \"0c12971f-da26-44a7-9333-1dba7e3f7be6\" to instance \"1af64c3e-264d-4ec1-ae7b-6199fb9bb475\" failed with: disk 0c12971f-da26-44a7-9333-1dba7e3f7be6 path /dev/vdb is attached to a different instance (12db4669-6145-449e-99b7-453031b71bb0)\n","stream":"stderr","time":"2017-12-26T23:21:55.189943108Z"}
{"log":"I1226 23:21:55.190080       1 actual_state_of_world.go:464] Report volume \"kubernetes.io/cinder/0c12971f-da26-44a7-9333-1dba7e3f7be6\" as attached to node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2017-12-26T23:21:55.190220094Z"}
{"log":"I1226 23:21:55.191217       1 event.go:218] Event(v1.ObjectReference{Kind:\"Pod\", Namespace:\"default\", Name:\"nginx-deployment-5fc77998d-gfws9\", UID:\"25d323df-ea92-11e7-bfcf-fa163ee4a110\", APIVersion:\"v1\", ResourceVersion:\"1198\", FieldPath:\"\"}): type: 'Warning' reason: 'FailedMount' AttachVolume.Attach failed for volume \"pvc-e46022b2-ea91-11e7-bfcf-fa163ee4a110\" : disk 0c12971f-da26-44a7-9333-1dba7e3f7be6 path /dev/vdb is attached to a different instance (12db4669-6145-449e-99b7-453031b71bb0)\n","stream":"stderr","time":"2017-12-26T23:21:55.191381995Z"}
{"log":"E1226 23:21:55.191456       1 nestedpendingoperations.go:263] Operation for \"\\\"kubernetes.io/cinder/0c12971f-da26-44a7-9333-1dba7e3f7be6\\\"\" failed. No retries permitted until 2017-12-26 23:23:57.190305974 +0000 UTC m=+1057.372010350 (durationBeforeRetry 2m2s). Error: \"AttachVolume.Attach failed for volume \\\"pvc-e46022b2-ea91-11e7-bfcf-fa163ee4a110\\\" (UniqueName: \\\"kubernetes.io/cinder/0c12971f-da26-44a7-9333-1dba7e3f7be6\\\") from node \\\"kube-dev-node-1-0\\\" : disk 0c12971f-da26-44a7-9333-1dba7e3f7be6 path /dev/vdb is attached to a different instance (12db4669-6145-449e-99b7-453031b71bb0)\"\n","stream":"stderr","time":"2017-12-26T23:21:55.191599561Z"}
{"log":"I1226 23:21:55.212179       1 actual_state_of_world.go:384] Set detach request time to current time for volume kubernetes.io/cinder/0c12971f-da26-44a7-9333-1dba7e3f7be6 on node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2017-12-26T23:21:55.212487697Z"}
{"log":"I1226 23:21:55.212270       1 node_status_updater.go:71] Could not update node status. Failed to find node \"kube-dev-node-1-1\" in NodeInformer cache. Error: 'node \"kube-dev-node-1-1\" not found'\n","stream":"stderr","time":"2017-12-26T23:21:55.212525699Z"}
```

jingxu97 commented:
@zetaab  Could you please confirm the behavior of cinder volumes? If the node instance is shundown, will it be deleted from the cloud provider? Also in this situation, is the volume still attached to the node even after the node is deleted from cloud provider?
AFAIK, other cloud providers such as GCE, AWS, if node is deleted from cloud provider, the volume should be detached automatically. If cinder is different, we have to solve the issue for it.

zetaab commented:
@jingxu97 check my answer in #57497

gnufied commented:
I think there are 2 different issues at play here:

1. First is - Cinder is not detaching volumes from shutdown instances even when pod that was using the volume is deleted. This is very reminiscent of bug we fixed in AWS recently (https://github.com/kubernetes/kubernetes/pull/55893). The underlying problem is - any code that makes a call to `GetVirtualMachineByName` will fail and throw error because a shutdown node is not included in the result.  The log line:

```
'node \"kube-dev-node-1-1\" not found'\n","stream":"stderr","time":"2017-12-26T23:21:55.212525699Z"}
```

exactly points to that. We need to make sure that - volume can be detached from a node even if it is shutdown. Code has to either handle "node not found" error or workaround it. I did something similar in AWS. 

2. Second is - looking at the logs, I think detach is failing from `node-1-1 ` because certain code path is trying to fetch the node and it can't(because not is shutdown) and hence whole detach operation fails. In a nutshell - it looks like we are trying detach but it isn't working because node is down.  The fix again has to make sure that, detach operation succeeds no matter node exists or not. I added some workaround for this stuff in AWS and similar workarounds has to be used for Cinder. 

@zetaab if you can find those workarounds and amend this PR that will be awesome. But I look this in some detail as well - I will have some free cycles in a day or two.

gnufied commented:
@jingxu97 I think the problem is similar to AWS. A shutdown actually still exists in Cinder(and hence has volumes attached to it) but some filtering rule or the other in Cinder cloudprovider is excluding such nodes. I haven't looked at Cinder cloudprovider code in detail to confirm, but that is my educated guess.

jingxu97 commented:
@gnufied thanks for the information. Could you provide the PR to fix AWS to amek sure detach succeed no matter node exist or not? Thanks!

zetaab commented:
after this fix:

```
% kubectl get nodes && kubectl get pv && kubectl get pods -o wide
NAME                  STATUS    ROLES     AGE       VERSION
kube-dev-master-1-0   Ready     <none>    3m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
kube-dev-node-1-0     Ready     <none>    1m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
kube-dev-node-1-1     Ready     <none>    1m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                  STORAGECLASS   REASON    AGE
pvc-372d2013-f0e5-11e7-987f-fa163eb81f6d   8Gi        RWO            Delete           Bound     default/nginxvolume1   zone1                    48s
NAME                               READY     STATUS    RESTARTS   AGE       IP          NODE
nginx-deployment-5fc77998d-9gcqj   1/1       Running   0          49s       10.36.0.3   kube-dev-node-1-1

% openstack server list && openstack volume list
+--------------------------------------+---------------------+--------+--------------------------------------------+
| ID                                   | Name                | Status | Networks                                   |
+--------------------------------------+---------------------+--------+--------------------------------------------+
| 5b0adfe1-6046-4769-a9b0-b062c9e13554 | kube-dev-proxy-1-0  | ACTIVE | kube-dev-net=192.168.18.74, 193.65.106.144 |
| d05f13d1-83ef-4406-92fa-b8b91bf4be85 | kube-dev-node-1-1   | ACTIVE | kube-dev-net=192.168.18.73                 |
| b7028a58-d15b-460e-af0f-fd5312a88912 | kube-dev-node-1-0   | ACTIVE | kube-dev-net=192.168.18.72                 |
| 8c1ab160-00cb-4a7a-a57e-67393632214f | kube-dev-master-1-0 | ACTIVE | kube-dev-net=192.168.18.71                 |
+--------------------------------------+---------------------+--------+--------------------------------------------+
+--------------------------------------+-------------------------------------------------------------+--------+------+--------------------------------------------+
| ID                                   | Display Name                                                | Status | Size | Attached to                                |
+--------------------------------------+-------------------------------------------------------------+--------+------+--------------------------------------------+
| 8290aef3-5507-40ac-aa2a-f41676ca2694 | kubernetes-dynamic-pvc-372d2013-f0e5-11e7-987f-fa163eb81f6d | in-use |    8 | Attached to kube-dev-node-1-1 on /dev/vdb  |
+--------------------------------------+-------------------------------------------------------------+--------+------+--------------------------------------------+

% openstack server stop d05f13d1-83ef-4406-92fa-b8b91bf4be85

% openstack server list
+--------------------------------------+---------------------+---------+--------------------------------------------+
| ID                                   | Name                | Status  | Networks                                   |
+--------------------------------------+---------------------+---------+--------------------------------------------+
| 5b0adfe1-6046-4769-a9b0-b062c9e13554 | kube-dev-proxy-1-0  | ACTIVE  | kube-dev-net=192.168.18.74, 193.65.106.144 |
| d05f13d1-83ef-4406-92fa-b8b91bf4be85 | kube-dev-node-1-1   | SHUTOFF | kube-dev-net=192.168.18.73                 |
| b7028a58-d15b-460e-af0f-fd5312a88912 | kube-dev-node-1-0   | ACTIVE  | kube-dev-net=192.168.18.72                 |
| 8c1ab160-00cb-4a7a-a57e-67393632214f | kube-dev-master-1-0 | ACTIVE  | kube-dev-net=192.168.18.71                 |
+--------------------------------------+---------------------+---------+--------------------------------------------+

here we need wait one - two minutes

% kubectl get pods -o wide
NAME                               READY     STATUS              RESTARTS   AGE       IP        NODE
nginx-deployment-5fc77998d-h2h6q   0/1       ContainerCreating   0          11s       <none>    kube-dev-node-1-0
% openstack volume list
+--------------------------------------+-------------------------------------------------------------+-----------+------+-------------+
| ID                                   | Display Name                                                | Status    | Size | Attached to |
+--------------------------------------+-------------------------------------------------------------+-----------+------+-------------+
| 8290aef3-5507-40ac-aa2a-f41676ca2694 | kubernetes-dynamic-pvc-372d2013-f0e5-11e7-987f-fa163eb81f6d | available |    8 |             |
+--------------------------------------+-------------------------------------------------------------+-----------+------+-------------+

It starts moving :)

% openstack volume list && kubectl get pods -o wide
+--------------------------------------+-------------------------------------------------------------+--------+------+--------------------------------------------+
| ID                                   | Display Name                                                | Status | Size | Attached to                                |
+--------------------------------------+-------------------------------------------------------------+--------+------+--------------------------------------------+
| 8290aef3-5507-40ac-aa2a-f41676ca2694 | kubernetes-dynamic-pvc-372d2013-f0e5-11e7-987f-fa163eb81f6d | in-use |    8 | Attached to kube-dev-node-1-0 on /dev/vdb  |
+--------------------------------------+-------------------------------------------------------------+--------+------+--------------------------------------------+
NAME                               READY     STATUS              RESTARTS   AGE       IP        NODE
nginx-deployment-5fc77998d-h2h6q   0/1       ContainerCreating   0          59s       <none>    kube-dev-node-1-0

% kubectl get pods -o wide
NAME                               READY     STATUS    RESTARTS   AGE       IP          NODE
nginx-deployment-5fc77998d-h2h6q   1/1       Running   0          1m        10.44.0.3   kube-dev-node-1-0
```

so now when node is shutdowned it will detach the drive. So this could be the fix for case 1) that @gnufied mentioned.

gnufied commented:
@jingxu97 the PR is https://github.com/kubernetes/kubernetes/pull/55893/files#diff-07ba008af9c76b0539556ff7fde3105eR2024 . What I ended up doing is, never make a `getInstance` call in code path that perform detach, because `getInstance` may itself fail and stop entire detach operation from proceeding. So rather than using instance information from `getInstance` I resorted to using `describeVolume` call to get as little of instance information I need - thereby avoiding need of full instance info. I think @zetaab is trying similar thing. 

@zetaab one thing to keep in mind is - we probably do not want to break existing behaviour for other stuff.

zetaab commented:
Well how we could avoid "breaking" current one?

gnufied commented:
@zetaab by breaking I meant - if current behaviour of openstack is that a shutdown node is not shown in node listing (via `oc get nodes` etc), then we should not change that behaviour for fixing this issue.

zetaab commented:
Ok, I am not changing that. Now that node behaviour is same, if node is shutdowned it cannot be seen in kubectl get nodes. The only difference is that if node gets deleted from cluster (which means that its state != "ACTIVE" in openstack), kubernetes will detach the cinder and raise volumedanglingerror which make it available for other nodes.

zetaab commented:
I retested and it works like should

```
% kubectl get pods -o wide
NAME                               READY     STATUS    RESTARTS   AGE       IP          NODE
nginx-deployment-5fc77998d-v875g   1/1       Running   0          51s       10.36.0.3   kube-dev-node-1-1
% kubectl get nodes
NAME                  STATUS    ROLES     AGE       VERSION
kube-dev-master-1-0   Ready     <none>    7m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
kube-dev-node-1-0     Ready     <none>    5m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
kube-dev-node-1-1     Ready     <none>    5m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
% openstack server list
openstack server stop +--------------------------------------+---------------------+--------+--------------------------------------------+
| ID                                   | Name                | Status | Networks                                   |
+--------------------------------------+---------------------+--------+--------------------------------------------+
| 126f8217-4959-4300-9605-7c9ad0a3e386 | kube-dev-proxy-1-0  | ACTIVE | kube-dev-net=192.168.18.78, 193.65.106.148 |
| 772d5aa3-1da2-4e50-8bd7-89ac552b5a2c | kube-dev-node-1-1   | ACTIVE | kube-dev-net=192.168.18.77                 |
| a3ff7989-f1b3-437a-b6a7-3f53fbce2e0f | kube-dev-node-1-0   | ACTIVE | kube-dev-net=192.168.18.76                 |
| b3be5da7-38c6-4d29-b518-83e81ebf6ede | kube-dev-master-1-0 | ACTIVE | kube-dev-net=192.168.18.75                 |
+--------------------------------------+---------------------+--------+--------------------------------------------+
% openstack server stop 772d5aa3-1da2-4e50-8bd7-89ac552b5a2c
% openstack server list
+--------------------------------------+---------------------+---------+--------------------------------------------+
| ID                                   | Name                | Status  | Networks                                   |
+--------------------------------------+---------------------+---------+--------------------------------------------+
| 126f8217-4959-4300-9605-7c9ad0a3e386 | kube-dev-proxy-1-0  | ACTIVE  | kube-dev-net=192.168.18.78, 193.65.106.148 |
| 772d5aa3-1da2-4e50-8bd7-89ac552b5a2c | kube-dev-node-1-1   | SHUTOFF | kube-dev-net=192.168.18.77                 |
| a3ff7989-f1b3-437a-b6a7-3f53fbce2e0f | kube-dev-node-1-0   | ACTIVE  | kube-dev-net=192.168.18.76                 |
| b3be5da7-38c6-4d29-b518-83e81ebf6ede | kube-dev-master-1-0 | ACTIVE  | kube-dev-net=192.168.18.75                 |
+--------------------------------------+---------------------+---------+--------------------------------------------+
% kubectl get nodes
NAME                  STATUS    ROLES     AGE       VERSION
kube-dev-master-1-0   Ready     <none>    8m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
kube-dev-node-1-0     Ready     <none>    6m        v1.8.0-alpha.2.7712+e59980c7882ed4-dirty
% kubectl get pods -o wide
NAME                               READY     STATUS    RESTARTS   AGE       IP          NODE
nginx-deployment-5fc77998d-27r6q   1/1       Running   0          2m        10.44.0.3   kube-dev-node-1-0
```

There you can see that when I shutdown node, it gets deleted from the kubernetes cluster.

However, I would like to see some solution to deal with pods which is in unknown state and pvc attached in it. See issues https://github.com/kubernetes/kubernetes/issues/51597 and https://github.com/kubernetes/kubernetes/issues/57643 That issue is common in all cloudproviders, so I am not going to make solution to that in this PR.

That "unknown state" pod+pvc issue is really important. Lets imagine situation that we have 2 nodes in some cloudprovider (it does not really matter which one). We have pod with pvc running in node1, then node1 goes to "notready" state for some reason. After that kubernetes will automatically try to evacuate the pod to node2. However, it cannot start the pod there because PVC is attached to node1. My proposal is that we could use cluster wide healthchecks to check pod status outside of that current node? Like if pod ip is 10.1.1.1, we could query that from any node and check is it answering or not. If it is not answering even those cluster wide healthchecks, we could just detach volume (then the pod is started in new node). It is not really good situation that if kubelet in node goes down for some reason, we cannot have control to any pod pvcs anymore.

jingxu97 commented:
@zetaab About when node in "unknown" state, our current suggestion is to have some outside controller to monitor it and delete the node from Kuberenetes API server. From kubernetes point of view, since node is "unknown", we can not be sure what is happening and make a decision to delete pod and detach volume, which might cause data corruption or data loss. If cluster admin/user monitors the node and decides to delete the node from kuberentes API server, that will solve the issue.

zetaab commented:
@jingxu97 that is actually quite cool solution! If node is in notready state more than x minutes -> execute kubectl delete node xx :)

zetaab commented:
@gnufied @dims can you check this PR and remove hold label if ok?

gnufied commented:
/hold cancel

gnufied commented on pkg/cloudprovider/providers/openstack/openstack.go:
```diff
@@ -346,11 +362,14 @@ func foreachServer(client *gophercloud.ServiceClient, opts servers.ListOptsBuild
-		Status: "ACTIVE",
+		Name: fmt.Sprintf("^%s$", regexp.QuoteMeta(mapNodeNameToServerName(name))),
 	}
+	if showOnlyActive {
```
In the interested of being backward compatible and keeping the diff small, `showOnlyActive` should default to true.  It is tricky to default a go boolean variable to true, so may be we can flip the variable name - `showAll`. What you think?

gnufied replied:
On second thoughts, it doesn't matter much. so I am going to lgtm.

gnufied commented:
/lgtm

gnufied commented:
/lgtm cancel

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -317,8 +319,33 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
+				return "", err
+			}
+			if srv.Status != "ACTIVE" {
+				err = os.DetachDisk(volume.AttachedServerId, volumeID)
```
Why are we performing detach in `attach` function? I thought - we made it clear that raising `NewDanglingError` should suffice

gnufied commented:
@zetaab the PR is still subtly wrong. There are two problems:

1. Did we handle the case of detaching volumes from shutdown nodes correctly? I do not think this PR handles that case properly. We need to modify `cinder.Detach` code path to make sure that `Detach` on a shutdown node works correctly.  Steps to reproduce:

- Create a pod (just a plain pod, not rc or deployment)
- Shutdown the node
- Delete the pod (pod actually will be deleted on its own after awhile anyways, but just to hasten the process)
- Make sure the volume that pod was using is detached correctly.


2. The second problem as I pointed above is - we should not perform `detach` in attach function. If `cinder.Detach` code path correctly - raising the Dangling error will be enough for detach to happen.

zetaab committed:
move detach out of os volumes attach

add test

add test

fix bazel

fix tests

change loglevel, remove else statement

zetaab commented:
@gnufied now it is like should be? However, the problem is that this is really slow compared to my first version of it. 

case 1) example (normal pod and node deleted)

{"log":"I0107 08:38:50.419673       1 attach_detach_controller.go:481] error removing node \"kube-dev-node-1-1\" from desired-state-of-world: failed to delete node \"kube-dev-node-1-1\" from list of nodes managed by attach/detach controller--the node still contains 1 volumes in its list of volumes to attach\n","stream":"stderr","time":"2018-01-07T08:38:50.419819439Z"}
{"log":"I0107 08:38:50.419743       1 actual_state_of_world.go:350] SetVolumeMountedByNode volume kubernetes.io/cinder/1571fadb-8d3d-46ee-b93b-9212856092a9 to the node \"kube-dev-node-1-1\" mounted true\n","stream":"stderr","time":"2018-01-07T08:38:50.419856206Z"}
{"log":"I0107 08:38:57.002578       1 actual_state_of_world.go:384] Set detach request time to current time for volume kubernetes.io/cinder/1571fadb-8d3d-46ee-b93b-9212856092a9 on node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2018-01-07T08:38:57.002723622Z"}
{"log":"W0107 08:44:57.036042       1 reconciler.go:233] attacherDetacher.DetachVolume started for volume \"pvc-e4805ac9-f385-11e7-9408-fa163eab910b\" (UniqueName: \"kubernetes.io/cinder/1571fadb-8d3d-46ee-b93b-9212856092a9\") on node \"kube-dev-node-1-1\" This volume is not safe to detach, but maxWaitForUnmountDuration 6m0s expired, force detaching\n","stream":"stderr","time":"2018-01-07T08:44:57.036215865Z"}
{"log":"I0107 08:45:05.948857       1 attacher.go:397] detached volume \"1571fadb-8d3d-46ee-b93b-9212856092a9\" from node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2018-01-07T08:45:05.949147141Z"}
{"log":"I0107 08:45:05.948908       1 operation_generator.go:417] DetachVolume.Detach succeeded for volume \"pvc-e4805ac9-f385-11e7-9408-fa163eab910b\" (UniqueName: \"kubernetes.io/cinder/1571fadb-8d3d-46ee-b93b-9212856092a9\") on node \"kube-dev-node-1-1\" \n","stream":"stderr","time":"2018-01-07T08:45:05.949197657Z"}

like you can see it goes to maxWaitForUnmountDuration before it will detach the volume. So it takes something like 6-8minutes to do that.

Then case 2) deployment with pvc and shutdown node where it is currently

I shutdowned node 08:52:30

{"log":"I0107 08:52:58.441782       1 attach_detach_controller.go:481] error removing node \"kube-dev-node-1-1\" from desired-state-of-world: failed to delete node \"kube-dev-node-1-1\" from list of nodes managed by attach/detach controller--the node still contains 1 volumes in its list of volumes to attach\n","stream":"stderr","time":"2018-01-07T08:52:58.4418363Z"}
{"log":"I0107 08:52:58.442010       1 actual_state_of_world.go:350] SetVolumeMountedByNode volume kubernetes.io/cinder/c573a507-5a60-47fa-a1f2-752058f3992b to the node \"kube-dev-node-1-1\" mounted true\n","stream":"stderr","time":"2018-01-07T08:52:58.442063636Z"}
{"log":"I0107 08:53:17.282033       1 actual_state_of_world.go:384] Set detach request time to current time for volume kubernetes.io/cinder/c573a507-5a60-47fa-a1f2-752058f3992b on node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2018-01-07T08:53:17.28234819Z"}
{"log":"W0107 08:59:17.349916       1 reconciler.go:233] attacherDetacher.DetachVolume started for volume \"pvc-ed55247e-f387-11e7-9408-fa163eab910b\" (UniqueName: \"kubernetes.io/cinder/c573a507-5a60-47fa-a1f2-752058f3992b\") on node \"kube-dev-node-1-1\" This volume is not safe to detach, but maxWaitForUnmountDuration 6m0s expired, force detaching\n","stream":"stderr","time":"2018-01-07T08:59:17.350035271Z"}
{"log":"I0107 08:59:19.678765       1 attacher.go:397] detached volume \"c573a507-5a60-47fa-a1f2-752058f3992b\" from node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2018-01-07T08:59:19.678985481Z"}
{"log":"I0107 08:59:19.678893       1 operation_generator.go:417] DetachVolume.Detach succeeded for volume \"pvc-ed55247e-f387-11e7-9408-fa163eab910b\" (UniqueName: \"kubernetes.io/cinder/c573a507-5a60-47fa-a1f2-752058f3992b\") on node \"kube-dev-node-1-1\" \n","stream":"stderr","time":"2018-01-07T08:59:19.67902802Z"}
{"log":"I0107 08:59:31.253076       1 actual_state_of_world.go:464] Report volume \"kubernetes.io/cinder/c573a507-5a60-47fa-a1f2-752058f3992b\" as attached to node \"kube-dev-node-1-0\"\n","stream":"stderr","time":"2018-01-07T08:59:31.253317782Z"}

and the pod was running in new node 09:01:45. So it took over 9 minutes to failover pod with pvc to another node. Is there any way to make this faster? Is it behaving like this in AWS? Also notable thing is that https://github.com/zetaab/kubernetes/blob/fixvolumeattached/pkg/cloudprovider/providers/openstack/openstack_volumes.go#L333 this row is not called at all in this case. At least I cannot see anything in controller logs.

zetaab commented:
I am pretty sure that the delay is here https://github.com/zetaab/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L193 this sets detach time to "current time" then it waits timeout https://github.com/zetaab/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L199 and then it uses force after timeout is true https://github.com/zetaab/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L228 here. We should have somekind of better way to detect this? @gnufied is this same problem in AWS or can I somehow fix this in pkg/volumes/cinder or in cloudprovider.

zetaab commented:
@jingxu97 I ping you because you are original author of this timeout thing https://github.com/kubernetes/kubernetes/blame/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L201 I am thinking that in case of node is shutdowned, we does not need to wait 6 minutes before we can detach the drive. Could we somehow query node status in somewhere and skip that timeout thing? What is your opinion? Currently ~9-10minutes of waiting in case of failover is too much. If we could get rid off that timeout thing, i think the time could be something like 1-3minutes.

jingxu97 commented:
@zetaab I understand your problem. Just to confirm that in case of node shutdown, do you plan to delete the node from the API server to trigger the pod deletion?

zetaab commented:
@jingxu97 yes, currently its working like "normally". However, I am thinking that how we could improve the speed of volume detachs if node is in "safe mode" to detach the volume. Like in case of openstack - if node is in shutdown state it is safe to detach the volume without waiting 6 minutes.

gnufied commented:
@zetaab thank you. this is looking much better. I will have deeper look.

One quick way to reproduce need for dangling error is:

1. Create a standalone pod (no deployments, rc etc) with volumes.
2. Shutdown the node.
3. Now wait for the pod on the node to be deleted. 
4. Once pod is deleted (spam `kubectl get pods`) but before controller-manager could detach the volume (that 6-8 time delay you noticed), restart the controller-manager
5. Above action will cause volume information to be wiped from controller-manager
6. Now try to attach same PVC in another pod (may be scheduled on different node). The pod will stuck in "ContainerCreating" state without the dangling volume fix.

zetaab commented:
```
{"log":"I0110 08:33:46.537006       1 openstack_volumes.go:333] volume 277592e1-b024-4cf1-a735-54896a1afb9f is already attached to node kube-dev-node-1-1 path /dev/vdb\n","stream":"stderr","time":"2018-01-10T08:33:46.537184701Z"}
{"log":"I0110 08:33:46.537045       1 attacher.go:159] Attach volume \"277592e1-b024-4cf1-a735-54896a1afb9f\" to instance \"6e426c80-a657-4068-9a79-794ca3cd574f\" failed with: disk 277592e1-b024-4cf1-a735-54896a1afb9f path /dev/vdb is attached to a different instance (0a86dbd4-68b3-4ddb-a0e5-ec8f752b8f7b)\n","stream":"stderr","time":"2018-01-10T08:33:46.537246501Z"}
{"log":"I0110 08:33:46.537138       1 actual_state_of_world.go:456] Add new node \"kube-dev-node-1-1\" to nodesToUpdateStatusFor\n","stream":"stderr","time":"2018-01-10T08:33:46.537253256Z"}
{"log":"I0110 08:33:46.537157       1 actual_state_of_world.go:464] Report volume \"kubernetes.io/cinder/277592e1-b024-4cf1-a735-54896a1afb9f\" as attached to node \"kube-dev-node-1-1\"\n","stream":"stderr","time":"2018-01-10T08:33:46.537258184Z"}
{"log":"E0110 08:33:46.537463       1 nestedpendingoperations.go:267] Operation for \"\\\"kubernetes.io/cinder/277592e1-b024-4cf1-a735-54896a1afb9f\\\"\" failed. No retries permitted until 2018-01-10 08:33:47.037360769 +0000 UTC m=+22.388373397 (durationBeforeRetry 500ms). Error: \"AttachVolume.Attach failed for volume \\\"pvc-7609193a-f5d7-11e7-acd2-fa163eda8e90\\\" (UniqueName: \\\"kubernetes.io/cinder/277592e1-b024-4cf1-a735-54896a1afb9f\\\") from node \\\"kube-dev-node-1-0\\\" : disk 277592e1-b024-4cf1-a735-54896a1afb9f path /dev/vdb is attached to a different instance (0a86dbd4-68b3-4ddb-a0e5-ec8f752b8f7b)\"\n","stream":"stderr","time":"2018-01-10T08:33:46.537595263Z"}
{"log":"I0110 08:33:46.537554       1 event.go:218] Event(v1.ObjectReference{Kind:\"Pod\", Namespace:\"default\", Name:\"nginx-deployment\", UID:\"f8e31fdd-f5e0-11e7-acd2-fa163eda8e90\", APIVersion:\"v1\", ResourceVersion:\"130530\", FieldPath:\"\"}): type: 'Warning' reason: 'FailedAttachVolume' AttachVolume.Attach failed for volume \"pvc-7609193a-f5d7-11e7-acd2-fa163eda8e90\" : disk 277592e1-b024-4cf1-a735-54896a1afb9f path /dev/vdb is attached to a different instance (0a86dbd4-68b3-4ddb-a0e5-ec8f752b8f7b)\n","stream":"stderr","time":"2018-01-10T08:33:46.537706402Z"}
```

then after 6 minutes 

```
{"log":"W0110 08:39:46.563829       1 reconciler.go:233] attacherDetacher.DetachVolume started for volume \"pvc-7609193a-f5d7-11e7-acd2-fa163eda8e90\" (UniqueName: \"kubernetes.io/cinder/277592e1-b024-4cf1-a735-54896a1afb9f\") on node \"kube-dev-node-1-1\" This volume is not safe to detach, but maxWaitForUnmountDuration 6m0s expired, force detaching\n","stream":"stderr","time":"2018-01-10T08:39:46.564021379Z"}
```

and then the pod starts in new node. I would say this works like should? Without working danlingerror this drive should not detach after controller restart, right?

zetaab commented:
@gnufied I tested this also with modifying openstack_volumes.go AttachDisk function to return just fmt.Errorf() instead of danglingvolumeerror. Then I reproduced steps where the danglingerror worked in last comment. There the danglingerror informed controller that it should detach drive first, and it worked. However, now with this fmt.errorf version it does not detach the drive at all and after 10minutes the drive is still there in shutdowned node. I think if I do not do anything, its still there in tomorrow. So from my point of view, I can say that this fixes at least those two issues that is mentioned in first post. Could you final check this PR and add LGTM if ok

gnufied commented:
@zetaab yep that is what we expected. Thank you for chasing this down and verifying it. I will lgtm it shortly.

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -586,6 +603,29 @@ func (os *OpenStack) DiskIsAttached(instanceID, volumeID string) (bool, error) {
+	instanceID := "/" + srv.ID
+	if ind := strings.LastIndex(instanceID, "/"); ind >= 0 {
+		instanceID = instanceID[(ind + 1):]
+	}
```
Can you add a comment about - why we are doing this? I assume that - instance-id of node exists after encountering first - "/" and it includes last "/" as well?  so something like - "foobar/instance-id/" ?

cc @rootfs @pospispa

zetaab replied:
Basically what that code does is that it removes first '/', but is that the case always? srv.ID usually in openstack is UID of the instance. So it should not contain anything more than UID, but I am not sure.

That code is taken originally from here https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cinder/attacher.go#L427-L439 I was thinking that I will use same code, if there is really some special cases.

zetaab replied:
git blame tells that this function is originally committed by @jamiehannaford Jamie could you help us, what is purpose of this? Are there some special cases why this is implemented like this?

zetaab replied:
might be that this is there because of this comment https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/openstack/openstack_instances.go#L161-L163 in future it is possible to return endpoint as: <endpoint>/<instanceid> and then this function will strip that <endpoint>/ out of it.

gnufied replied:
cool. thank you for digging this. I was wondering about it a bit..

jamiehannaford replied:
@zetaab Pretty sure I did not add this. Not sure why blame is saying otherwise. Looks like that comment addresses the reason though.

gnufied commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -600,6 +640,32 @@ func (os *OpenStack) DisksAreAttached(instanceID string, volumeIDs []string) (ma
+	}
+	return os.DisksAreAttached(instanceID, volumeIDs)
+}
+
```
So this change is unrelated to issues this PR is supposed to fix right? I think this is fixing `VolumesAreAttached` mechanism but can we extract that into a separate PR? I feel as it is - we are fixing 2 bugs in this PR. This will be 3rd one.

zetaab replied:
hmm, imo this should be here. Without this, if you shutdown node, it will mark all volumes as "available", but those are really attached to shutdowned node. This PR should fix those detach problems if node is shutdowned.

gnufied reviewed (changes requested):
Mostly looks good. Some minor comments.

zetaab commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -578,6 +592,9 @@ func (os *OpenStack) GetAttachmentDiskPath(instanceID, volumeID string) (string,
 func (os *OpenStack) DiskIsAttached(instanceID, volumeID string) (bool, error) {
+	if instanceID == "" {
+		glog.Warningf("calling DiskIsAttached with empty instanceid: %s %s", instanceID, volumeID)
+	}
```
well there was case where the code did not work properly if instanceid was "". I can remove it, but I was thinking that it could help debugging in future.

gnufied commented:
Please squash last commit and then I think we are good to go.

zetaab commented:
done

gnufied commented:
/lgtm

gnufied commented:
cc @anguslees  for approval from openstack side of things.

zetaab commented:
/retest

FengyunPan commented:
@gnufied @zetaab Thanks.
/approve

jianglingxia commented:
the pr can merge v1.8 release? I use v1.8.5 version and test shutdown one minion have the same problem#57497,thanks for you reply!
